### ISSUE TITLE
"Preformatted" block autocompleting type declarations in pasted code

### ISSUE THREAD

**Issue by koskila** [WordPress/gutenberg#14092]
**Describe the bug**
"Preformatted" block is "autocompleting" or "autoclosing" type declarations in pasted code, because it interprets them as html tags.

**To Reproduce**
1. Edit page
2. Add Preformatted block
3. Paste code, such as "DbContextOptions<ApplicationDbContext> options"
4. Modify the code as html
5. Save
6. Observe the "preformatted" block now having </ApplicationDbContext> at the end.

**Expected behavior**
Type declarations should not be autoclosed - they are independent, and not html.

**Screenshots**
![image](https://user-images.githubusercontent.com/4086860/53357365-bc620d00-38fb-11e9-95ed-2b771a83ab22.png)
(Example with <int> being closed at the end of the file.

**Desktop (please complete the following information):**
 - Browser: Chrome Version 72.0.3626.109 (Official Build) (64-bit), but I suspect this might be partially caused by mobile usage as that's when I usually see this
 - WordPress 5.1

**Smartphone (please complete the following information):**
 - Device: Galaxy S9+
 - OS: Android 9 (4.9.112-15119402)
 - Browser Chrome
 - Version 72.0.3626.96

Let me know if you require any more information!

**Comment by talldan** [comment]
Hi @koskila, I had a quick test of this but wasn't able to reproduce the issue. Would be able to provide the snippet of code you're using to trigger the issue? Thanks.

I also wanted to check, when you mentioned 'Modify the code as html', were you using the option 'Edit as HTML' in the block menu? I was also wondering why you'd need to do that?

**Comment by koskila** [comment]
@talldan - I need to edit blocks as html to add a parameter lang="" for the preformatted blocks, as there's no out-of-the-box block providing syntax highlighting. Modifying the blocks selecting "Edit as HTML" brings this issue up - not always, but often:
![image](https://user-images.githubusercontent.com/4086860/54574859-55021f00-49c8-11e9-8fee-d7b042016ceb.png)

Removing the "closing tags" from the end sometimes is saved, sometimes not - I have no idea why it's not consistent...

**Comment by koskila** [comment]
Okay, so I tried again. Adding this to a block using "Edit as HTML":

<pre lang="csharp">
services.AddSingleton<IHttpContextAccessor, HttpContextAccessor>();
</pre>

Produces this:
![image](https://user-images.githubusercontent.com/4086860/54575856-34d45f00-49cc-11e9-83d3-d9cff9a169ba.png)

**Comment by koskila** [comment]
Alright, another example - I have this in the editor:
![image](https://user-images.githubusercontent.com/4086860/54799380-52efc880-4c33-11e9-851f-a709ba0f6e99.png)

When I hit "Preview", this is what I get:
![image](https://user-images.githubusercontent.com/4086860/54799391-600cb780-4c33-11e9-91f9-c9ddf368fabc.png)

This didn't happen before Gutenberg, but of course that doesn't necessarily mean it's caused by it.

**Comment by koskila** [comment]
Actually, scratch that. After opening the page in edit mode in another browser with WP-GeSHi-Highlight disabled, the issue is there again. So it's probably not that plugin, but rather Gutenberg.

Sorry for the back-and-forth. What an infuriating case. If you want any more info, let me know - would love to see this fixed one way or another.

![image](https://user-images.githubusercontent.com/4086860/54799709-b8908480-4c34-11e9-9be9-969aa539f9b5.png)

**Comment by ellatrix** [comment]
@koskila The preformatted block is _not_ a code editor. You should use the Code block instead. In the future, the code block have an option to set the language and have syntax highlighting.

**Comment by koskila** [comment]
Thanks for the comments - it's not the preformatted block though, it's the custom html block (like shown in the screenshots). Even a way to disable the autocomplete (for custom html blocks) from messing up would be a perfectly valid solution.

**Comment by ellatrix** [comment]
Sorry, perhaps I'm having a hard time understanding the issue.

**Comment by koskila** [comment]
No worries - so am I! But my understanding is, that whenever I want to have something like this in a **custom HTML block**:
```
<pre>
var list = new List<string>();
</pre>
```

It'll get turned to this:
```
<pre>
var list = new List<string></string>();
</pre>
```

Or occasionally this:
```
<pre>
var list = new List<string>();
</string>
</pre>
```

And it gets even weirder with Dictionaries and such:

```
<pre>
var dict = new Dictionary<string,Object>();
</pre>
```

Will become:
```
<pre>
var dict = new Dictionary<string ='',object =''>();
</string></object>
</pre>
```
(Note: even casing changes!)

**Comment by koskila** [comment]
Oh, and as a side note - really looking forward to the changes to code-block you mentioned! Syntax highlighting is a very welcome new feature. Much appreciated :)

**Comment by talldan** [comment]
@koskila Right, I understand now. The issue is that the Custom HTML block does some HTML validation and automatic fixing when it sees invalid html. When it sees `<string>` it considers it an unterminated html tag and so adds a closing tag. It'll also changing the casing as you mentioned.

The solution (if you want to use the Custom HTML block instead of the Code block) is to escape the code you're trying to add first. Basically turn this:
```
<pre>
var list = new List<string>();
</pre>
```

into this:
```
<pre>
var list = new List&lt;string&gt;();
</pre>
```
There are some online tools you can use to escape content, like this one: https://www.freeformatter.com/html-escape.html

This tool might also be useful - http://hilite.me/ - it converts code into syntax highlighted html snippet that you can insert into your post (and does the escaping as part of that). 

I would just double-check to make sure there are no malicious scripts or anything like that are being added to the HTML! (not sure, but Gutenberg might strip those out anyway).

I'll close the issue now as I think there are some good solutions, and hopefully at some point syntax highlighting will appear in the code block.

**Comment by koskila** [comment]
@talldan, I appreciate the response, but encoding the html does not help. See - this:
![image](https://user-images.githubusercontent.com/4086860/57198782-9bfba200-6f7f-11e9-9339-d47f49f40274.png)

Becomes this:
![image](https://user-images.githubusercontent.com/4086860/57198785-b766ad00-6f7f-11e9-990b-021f5d0f9b0e.png)

I'd love to see the syntax highlighting appear, as this renders Gutenberg really difficult to use for sharing code snippets. I could upload everything to GitHub as Gists and then embed them, but that's a really overengineered problem of pasting example code - sometimes just one line - to a blog entry.

I'd like for the issue to be reopened as there's no good workaround (apart from not using Gutenberg but embedding the scripts instead) available, please.

**Comment by talldan** [comment]
@koskila - please could you provide reproduction steps and let us know which version you're using? I'm unable to reproduce the issue. Here's an example of what I see with the custom HTML block:

**Editor - Custom HTML block content**
![Screen Shot 2019-05-06 at 1 00 06 pm](https://user-images.githubusercontent.com/677833/57207300-20413a00-6fff-11e9-9a34-ff3da06e92e1.png)

**Editor - Custom HTML block preview**
![Screen Shot 2019-05-06 at 1 00 13 pm](https://user-images.githubusercontent.com/677833/57207306-2b946580-6fff-11e9-806c-4d256bf6fe5a.png)

**Post preview**
![Screen Shot 2019-05-06 at 1 00 20 pm](https://user-images.githubusercontent.com/677833/57207308-33eca080-6fff-11e9-9a37-093bf0b924de.png)

**Comment by talldan** [comment]
As for reopening—if we can find a bug with the Custom HTML block, then I'm happy to open this as a bug. At the moment I'm unable to reproduce an issue. We'll need some reproduction steps in the latest version of Gutenberg (it could be that there was an issue in an older version of the editor that's since been fixed).

There's already a separate open issue for syntax highlighting in the code block that you can follow - https://github.com/WordPress/gutenberg/issues/10423/

There isn't a designated developer for the issue yet, but there's already been some discussion and a few proof of concepts.

**Comment by talldan** [comment]
I still couldn't reproduce this using the same text in your post. I think there are two possibilities:
- This is a bug that's since been fixed (WordPress 5.2 was released yesterday, so there's a chance that might resolve the matter) 🤞 
- The theme or a plugin are causing an issue

If it can be isolated to a theme or plugin issue, it would be worth following up with the developer. I'm happy to test with the theme you're using if it's freely available.
